### ISSUE TITLE
False positive with Marshaler/Umarshaler interfaces and other methods

### ISSUE THREAD

**Issue by hhromic** [raeperd/recvcheck#14]
As of version `v0.2.0`, I'm getting a false positive from `recvcheck` when using the `Marshaler` and `Unmarshaler` interfaces together with other methods (consistently using receivers) in the same struct.

Consider the following minimal reproducible example (MRE):
```go
// Package main provides main application code.
package main

import "errors"

// SomeType is some type.
type SomeType struct{}

// String returns the string representation.
func (t SomeType) String() string {
	return "SomeType"
}

// MarshalText implements [encoding.TextMarshaler] by calling [SomeType.String].
func (t SomeType) MarshalText() ([]byte, error) {
	return []byte(t.String()), nil
}

// UnmarshalText implements [encoding.TextUnmarshaler].
// It accepts any slice of bytes produced by [SomeType.MarshalText].
func (t *SomeType) UnmarshalText(b []byte) error {
	str := string(b)
	if str == "SomeType" {
		*t = SomeType{}
		return nil
	}

	return errors.New("unknown text for unmarshaling")
}

func main() {
}
```

In this example, I'm implementing both, the `Stringer` interface and the `Marshaler`/`Unmarshaler` interface pair.

When I run `golangci-lint` with the following minimal configuration:
```yaml
version: '2'

linters:
  enable:
    - recvcheck
```

I get the following linter error:
```
$ golangci-lint run
main.go:7:6: the methods of "SomeType" use pointer receiver and non-pointer receiver. (recvcheck)
type SomeType struct{}
     ^
1 issues:
* recvcheck: 1
```

I think this is a false positive, because if we are to ignore the (Un)Marshaler interfaces, then the only method left is `String()` which, due to being the only method, should not be considered "inconsistent" in its receiver.

I noticed that if the only method apart from the (Un)Marshaler interfaces is a pointer-receiver, then the linter is happy. This leads me to think that the linter is only ignoring `Marshal*` interfaces and not their potential `Unmarshal*` counterpart ones.

**Comment by ldez** [comment]
Hello,

`MarshalText` is excluded by default, but `UnmarshalText` is not. This is intentional, so it's not a false positive.

You can exclude `UnmarshalText` if you want:

```yml
version: '2'

linters:
  enable:
    - recvcheck
  settings:
    recvcheck:
      exclusions:
        - '*.UnmarshalText'
```

**Comment by hhromic** [comment]
@ldez thanks for your quick reply, explanation and workaround for my case. It is much appreciated.

> `MarshalText` is excluded by default, but `UnmarshalText` is not. This is intentional, so it's not a false positive.

This is interesting. If you have some time I would love to read the reasoning behind this decision.

By not excluding `Unmarshal*` functions as well as `Marshal*`, you essentially are forcing consistency towards using pointer receivers, which is not always the right assumption. For example in my initial MRE where I'm combining a Stringer with a Marshaler/Unmarshaler, which I think is a perfectly valid use case. In this case, the consistency should be towards non-pointer receivers but the linter, by default, will complain until you configure it to ignore the situation.

This is why I initially thought this is a false-positive. False, because I don't see why this code is not good practice.

**Comment by ldez** [comment]
> However, what happens when you are not in doubt, like in the MRE of this issue?

The linter reports (it's the same thing for all linters) should always be considered in this context, and the relevance should be evaluated by a human; this is why there is configuration.

This linter has started as a linter to enforce pointer usage.
I made some suggestions to enforce consistency over pointer only, but the linter still has in this "ADN" the pointer priority.
In all cases, a priority should be defined, and I think pointer priority is right (based on the recommendations).

Marshaller and unmarshaller methods have different constraints. There are discussions inside the issues about this topic.
